### ISSUE TITLE
Pinned workspaces

### ISSUE THREAD

**Issue by maria-komarova** [pop-os/cosmic-workspaces-epoch#108]
Pinned workspaces is a new feature meant as an alternative to the fixed (or static) workspaces. The functionality in essence is very similar. The main difference is that workspaces are pinned in the window and workspaces overview instead of one using "Add workspace" button after switching to fixed workspaces in Settings. Having pinned workspace should allow us to avoid having a separate mode in the implementation. 
- Pin button is in the right bottom corner of each workspace thumbnail (preview) in the workspace picker and shows on hover (it should be visible in default state on touch devices). Once that button is clicked, it is toggled on. 
- The workspaces that are pinned are not removed even when empty unlike other dynamic workspaces. 
- There is always an unpinned empty dynamic workspace at the end. 

Mockup - Workspace 1 is pinned, Workspace 3 is hovered.
![workspace1-overview-hoverW3](https://github.com/user-attachments/assets/90bd7fbd-27bb-4a13-9b07-3fb77f53302c)

If the window (Text Editor) on workspace 2 is closed on the mockup below, then one navigates away, workspace 2 is deleted because it is unpinned and therefore works as a dynamic workspace. 
![workspace2-overview-pinnedW1W3](https://github.com/user-attachments/assets/e8da947b-d16f-493f-95f2-e40320f50606)

[Figma prototype](https://www.figma.com/proto/Co13PwCHl1Num6VV2GnUii/Workspaces-First-Release-Developer-Handoff?node-id=4-13793&p=f&t=6W8tglYloy4Cg6Dp-1&scaling=contain&content-scaling=fixed&page-id=1%3A6&starting-point-node-id=4%3A13793):
- Open overview with the mouse
- Pin workspace 1 and workspace 3
- Close Text Editor on workspace 2 from inside the overview
- Navigate to workspace 1 or 3 from workspace 2 
- Close Files browser on workspace 1
- Navigate away from workspace 1.

[Figma mockups](https://www.figma.com/design/Co13PwCHl1Num6VV2GnUii/Workspaces-First-Release-Developer-Handoff?node-id=1-6&t=2xSt9qhmDluf8T4H-1).

**Comment by maria-komarova** [comment]
I'm not sure what you mean by the automated placement of a set number of pinned workspaces. 
I don't know if this might partially answer your question but while workspaces are created dynamically by default, pinning the workspaces from the overview should essentially turn them into static ones since we plan for pinned workspaces to persist through reboots.

**Comment by joedavid0** [comment]
Personally, the issue I have is not necessarily that COSMIC "removes" the workspace but rather that it 
1. prohibits one to manually create workspaces at any number and
2. upon a workspace becoming empty, moves all higher workspaces one number lower.
I always think about workspaces being tied to a hotkey I have the press but (2) causes workspaces to always shift around (see also pop-os/cosmic-comp#586). The problem with (1) is that I often have the same application on the same workspace (and thus hotkey) but when I e.g. want to put an app on number 5, I have to create 2, 3 and 4 first.

**Comment by theshatterstone** [comment]
> I'm not sure what you mean by the automated placement of a set number of pinned workspaces.

@maria-komarova I'm sorry, what I meant was if there will be an exposed way to create scripts that would create a set number of fixed workspaces automatically. An exposed command line tool or config file that we could modify via scripts.

**Comment by tarruda** [comment]
> > I'm not sure what you mean by the automated placement of a set number of pinned workspaces.
> 
> [@maria-komarova](https://github.com/maria-komarova) I'm sorry, what I meant was if there will be an exposed way to create scripts that would create a set number of fixed workspaces automatically. An exposed command line tool or config file that we could modify via scripts.

Seems like this is implemented as a request at cosmic protocol: https://github.com/pop-os/cosmic-workspaces-epoch/pull/128/files

Not very familiar with wayland, but I'm assuming it should be possible to write a wayland/cosmic client/script that can do the pinning automatically

**Comment by theshatterstone** [comment]
@tarruda Yeah, I was considering this because I thought it might have been necessary but it seems like it won't be needed for my usecase. I didn't delete my question because I thought a response might be useful to other people with similar questions

**Comment by Etienne-TR** [comment]
For me, a simple solution would be to add two options in the dynamic workspace settings:
1. Minimum number of workspaces
2. Keep empty workspace if not last

Some arguments for these features:
* Having a minimum number of workspaces would provide consistency for users who prefer a fixed workspace structure while still allowing dynamic expansion
* The "Keep empty workspace if not last" option would prevent accidental workspace deletion when temporarily moving all windows from a middle workspace, preserving your organizational system
* These options would balance the flexibility of dynamic workspaces with the predictability of fixed workspaces
* Users with specific workflow patterns could maintain their preferred workspace layout even during less busy periods, preventing disruption to established habits and mental organization.

**Comment by Etienne-TR** [comment]
> [@Etienne-TR](https://github.com/Etienne-TR) Considering existing window managers either have dynamic or static workpaces (or an option to toggle between the 2, such as gnome 3), I wonder how many users would prefer fixed + dynamic expansion.

For users,  it appears to be only dynamic, as it is now. It's simple and refers to something familiar. (For developers, there's no need to implement a specific static behavior, but only to adjust the dynamic behavior with parameters.)

For those used to static, the two proposed options to adjust dynamic behavior address different use cases in an explicit way. I think this could be easily adopted by users from other desktop environments.

**Comment by SergejGorev** [comment]
> For me, a simple solution would be to add two options in the dynamic workspace settings:
> 
> 1. Minimum number of workspaces
> 2. Keep empty workspace if not last
> 
> Some arguments for these features:
> 
> * Having a minimum number of workspaces would provide consistency for users who prefer a fixed workspace structure while still allowing dynamic expansion
> * The "Keep empty workspace if not last" option would prevent accidental workspace deletion when temporarily moving all windows from a middle workspace, preserving your organizational system
> * These options would balance the flexibility of dynamic workspaces with the predictability of fixed workspaces
> * Users with specific workflow patterns could maintain their preferred workspace layout even during less busy periods, preventing disruption to established habits and mental organization.

This seems to me like the best compromise. IMHO the whole point of having multiple workspaces and being able to navigate them via a shortcut is to access same applications on the same workspace over and over again. Cuz its convenient to switch around that way. You always remember for example that worksapce1=browser, workspace2=terminals, worspace3=someOtherApp. I think linux community will be thankful for this feature. It is a vital necessity for me personally, I cant imagine having smth else. And this is the number one reason why I struggle to switch to cosmic yet.

**Comment by noxxxxxious** [comment]
> > Having a minimum number of workspaces would provide consistency for users who prefer a fixed workspace structure while still allowing dynamic expansion
> 
> [@Etienne-TR](https://github.com/Etienne-TR) Considering existing window managers either have dynamic or static workpaces (or an option to toggle between the 2, such as gnome 3), I wonder how many users would prefer fixed + dynamic expansion.

Read your comment when coming here to check on the status of fixed workspaces, but figured I'd weigh in with me being in the camp of "fixed only", though I'm sure others will prefer dynamic expansion as well. 

Maybe this is a Gnome bug, maybe it's a Gnome feature, but in Gnome, when you have multiple monitors and a set a fixed number of workspaces, if you do anything that causes your monitor numbers to drop (like disconnecting from a dock), Gnome likes to do something along the lines of treat each individual monitor as its own workspace, so even though I have my workspaces set to a fixed number of 5, if I unplug my laptop and open it up while on the go, even in the settings menu, it will be changed to something like 13 or whatever, depending on how many windows I had open across different monitors on different workspaces.

If I, personally, could always have it just be 5, I would.
